### PR TITLE
Feature: cibadmin: render access mode for selected user

### PR DESCRIPTION
**Pull request by jnpkrn**
Known rough edges:

* [x] fix oddities spotted by Ken
* [x] missing deployment stuff (Makefile/spec) for new `acls-render.xsl`
* [x] cibadmin's help should show an example usage
* [x] not tested user friendliness under all circumstances (`-U` required etc.)
* [x] no CLI tests
* [ ] public API not reflecting the intent to generalize for groups alike (worked on)
* [x] check that we are working with eligible schema (to begin with)

### CODE REVIEW

**Comment by jnpkrn** [comment]
and actually meant to take a look at irking sync/async handling redundancy
for output (to this point, was only testing with sync/`-s`)

File: include/crm/common/xml.h [hunk 1]
```diff
@@ -301,6 +301,29 @@ int xml_apply_patchset(xmlNode *xml, xmlNode *patchset, bool check_version);
 
 void patchset_process_digest(xmlNode *patch, xmlNode *source, xmlNode *target, bool with_digest);
 
+enum acl_ns {
+    acl_ns_writable = 0x01,
+    acl_ns_readable = 0x02,
+    acl_ns_denied = 0x04,
+};
```
Comment (kgaillot) [inline include/crm/common/xml.h]: Wouldn't it make sense to reuse the xpf_acl flags?

File: lib/common/acl.c [hunk 2]
```diff
@@ -720,4 +727,331 @@ crm_acl_get_set_user(xmlNode *request, const char *field, const char *peer_user)
...
+    return ret;
+}
+
+int
+pcmk__acl_evaled_render(xmlDoc *annotated_doc, enum pcmk__acl_render_how how,
+                        xmlChar **doc_txt_ptr, int *doc_txt_len)
+{
+    xmlDoc *xslt_doc;
+    xsltStylesheet *xslt;
+    xsltTransformContext *xslt_ctxt;
```
Comment (kgaillot) [inline lib/common/acl.c]: Don't we have to guard everything with HAVE_LIBXSLT?

File: xml/acl-render.xsl [hunk 3]
```diff
@@ -0,0 +1,225 @@
+<!--
+ Copyright 2019 the Pacemaker project contributors
+
+ The version control history for this file may have further details.
+
+ Licensed under the GNU General Public License version 2 or later (GPLv2+).
+ -->
```
Comment (kgaillot) [inline xml/acl-render.xsl]: As we start getting more unrelated components under xml, I'd like to start using subdirectories for logical purposes (eventually I'd like to move all the CIB XML RNGs/XSLTs to xml/cib). Let's put this in a new subdirectory.

File: lib/common/xml.c [hunk 4]
```diff
@@ -3174,21 +3171,22 @@ crm_xml_dump(xmlNode * data, int options, char **buffer, int *offset, int *max,
          */
         xmlBufferSetAllocationScheme(xml_buffer, XML_BUFFER_ALLOC_DOUBLEIT);
 
-        *max = xmlNodeDump(xml_buffer, doc, data, 0, (options & xml_log_option_formatted));
-        if (*max > 0) {
-            *buffer = strdup((char *)xml_buffer->content);
+        ret = xmlNodeDump(xml_buffer, doc, data, 0, (options & xml_log_option_formatted));
```
[the remark this answers is not in the document]
Comment (jnpkrn) [reply lib/common/xml.c]: It's actually next to impossible unless one wants to fiddle with
struct fields and make some assumptions about the invariants, since
`xmlBufCreate` is a private API.  The respective code comment advice
that you reproduced here is likely wrong (or comes from when it was
undecided or something):

https://mail.gnome.org/archives/xml/2019-July/msg00009.html

File: lib/common/xml.c [hunk 5]
```diff
@@ -3174,21 +3171,22 @@ crm_xml_dump(xmlNode * data, int options, char **buffer, int *offset, int *max,
...
         if ((now + 1) < next) {
             crm_log_xml_trace(data, "Long time");
             crm_err("xmlNodeDump() -> %dbytes took %ds", *max, next - now);
         }
+#endif
 
         xmlBufferFree(xml_buffer);
         return;
     }
-#endif
```
[the remark this answers is not in the document]
Comment (jnpkrn) [reply lib/common/xml.c]: I think it could go even smoother, since libxml supports IO without
externally visible buffers.

[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/xml.c]: Great, that would be even cleaner

File: include/crm/common/xml.h (hunk 1 above)
Comment (jnpkrn) [reply include/crm/common/xml.h]: Could be, but at the cost of publicizing them, which may not be a good
move :-/

Comment (kgaillot) [reply include/crm/common/xml.h]: Oh right, never mind

**Comment by jnpkrn** [comment]
2.5 (2.9) points from the initial list is still somewhat missing,
but shall work better now, and be close to the final version.

File: lib/common/xml.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (jnpkrn) [reply lib/common/xml.c]: Included now.

File: lib/common/acl.c (hunk 2 above)
Comment (jnpkrn) [reply lib/common/acl.c]: Thanks, shall be fixed now.

File: lib/common/xml.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (jnpkrn) [reply lib/common/xml.c]: See the first commit.

File: xml/acl-render.xsl (hunk 3 above)
Comment (jnpkrn) [reply xml/acl-render.xsl]: Resolved with enumerated "namespaces" as mentioned via other channel.

File: lib/common/xml.c [hunk 6]
```diff
@@ -3137,30 +3139,25 @@ crm_xml_dump(xmlNode * data, int options, char **buffer, int *offset, int *max,
...
         return;
     }
-#if 0
-    if (is_not_set(options, xml_log_option_filtered)) {
-        /* Turning this code on also changes the scheduler tests for some reason
-         * (not just newlines).  Figure out why before considering to
-         * enable this permanently.
-         *
-         * It exists to help debug slowness in xmlNodeDump() and
-         * potentially if we ever want to go back to it.
```
[the remark this answers is not in the document]
Comment (jnpkrn) [reply lib/common/xml.c]: Thanks, that was useful.

Partially resolv{ed,able} with ditching the idea of strictly holistic
accummulate-then-dump that appears deeply stuck at all costs up until
now, not sure why -- it looks like textbook influence, perhaps, but
regarding IO efficiency, I wouldn't expect doing any worse with what's
basically a series of `write` as opposed to series of `brk` and `mmap`
that `realloc` amounts to (context switches) and there actually shall
not be frequent page faults in the IO case.

And when any further optimizations are due, it's possible to devise
a separate serialization function using libc IO functions with their
own buffering instead...

File: lib/common/xml.c [hunk 7]
```diff
@@ -3174,21 +3171,22 @@ crm_xml_dump(xmlNode * data, int options, char **buffer, int *offset, int *max,
          */
         xmlBufferSetAllocationScheme(xml_buffer, XML_BUFFER_ALLOC_DOUBLEIT);
 
-        *max = xmlNodeDump(xml_buffer, doc, data, 0, (options & xml_log_option_formatted));
-        if (*max > 0) {
-            *buffer = strdup((char *)xml_buffer->content);
+        ret = xmlNodeDump(xml_buffer, doc, data, 0, (options & xml_log_option_formatted));
+        if (ret > 0) {
+            buffer_print(*buffer, *max, *offset, "%s", (char *)xml_buffer->content);
```
[the remark this answers is not in the document]
Comment (jnpkrn) [reply lib/common/xml.c]: Hmm, it was a blind faith on my side, I admit, I shall look into it more
I guess.

File: lib/common/xml.c [hunk 8]
```diff
@@ -3290,6 +3298,17 @@ dump_xml_unformatted(xmlNode * an_xml_node)
...
 
+void
+pcmk__xml_serialize_fd_formatted(int fd, xmlNode *cur)
+{
+    xmlOutputBuffer *fd_out = xmlOutputBufferCreateFd(fd, NULL);
+    CRM_ASSERT(fd_out != NULL);
+    xmlNodeDumpOutput(fd_out, cur->doc, cur, 0, 1, NULL);
+    xmlOutputBufferWrite(fd_out, sizeof("\n") - 1, "\n");  /* final NL */
+    xmlOutputBufferClose(fd_out);
+    fdatasync(fd);
```
Comment (kgaillot) [inline lib/common/xml.c]: Apparently fdatasync() can be buggy in FreeBSD (and isn't available before FreeBSD 11.1). Since meta-data has to be sync'd due to file size change, there won't be any benefit over fsync() anyway.

File: include/crm/common/xml_internal.h [hunk 9]
```diff
@@ -142,4 +142,34 @@ do {
  */
 void pcmk__xml_serialize_fd_formatted(int fd, xmlNode *cur);
 
+enum pcmk__xml_artefact_ns {
+    pcmk__xml_artefact_ns_legacy_rng = 1,
+    pcmk__xml_artefact_ns_legacy_xslt,
+    pcmk__xml_artefact_ns_base_rng,
+    pcmk__xml_artefact_ns_base_xslt,
+};
```
Comment (kgaillot) [inline include/crm/common/xml_internal.h]: spelling: artifact

File: include/crm/common/xml.h [hunk 10]
```diff
@@ -312,6 +312,48 @@ int xml_apply_patchset(xmlNode *xml, xmlNode *patchset, bool check_version);
 
 void patchset_process_digest(xmlNode *patch, xmlNode *source, xmlNode *target, bool with_digest);
 
+/* exclusive (should any combination make sense -> explicitly enumerated),
+   with intentional symbolic constant duality (easier to adapt to growth) */
+enum pcmk_acl_cred_type {
+#define PCMK_ACL_CRED_USER PCMK_ACL_CRED_USER
```
Comment (kgaillot) [inline include/crm/common/xml.h]: I don't understand what this definition accomplishes

File: include/crm/common/xml.h [hunk 11]
```diff
@@ -312,6 +312,48 @@ int xml_apply_patchset(xmlNode *xml, xmlNode *patchset, bool check_version);
...
+ * \note Only supported schemas are those following acls-2.0.rng, that is,
+ *       those validated with pacemaker-2.0.rng and newer (artificially caped
+ *       at 4.0, not including it (the future cannot be predicted,
+ *       compatibility needs to be confirmed, then, hopefully leading to
+ *       this comment being updated).
+ */
+int pcmk_acl_evaled_as_namespaces(enum pcmk_acl_cred_type cred_type,
+                                  const char *cred, xmlDoc *cib_doc,
+                                  xmlDoc **acl_evaled_doc);
+
```
Comment (kgaillot) [inline include/crm/common/xml.h]: Let's make all of this internal API, at least for now. I'd rather keep the ability to change it freely until we've had some experience with it.

File: lib/common/acl.c [hunk 12]
```diff
@@ -798,4 +805,342 @@ crm_acl_get_set_user(xmlNode *request, const char *field, const char *peer_user)
...
+    }
+
+    /* XXX see the comment for this function, pacemaker-4.0 may need
+           updating respectively in the future */
+    validation = crm_element_value(xmlDocGetRootElement(cib_doc),
+                                   XML_ATTR_VALIDATION);
+    version = get_schema_version(validation);
+    if (get_schema_version("pacemaker-2.0") > version
+            || (-1 != get_schema_version("pacemaker-4.0")
+                    && get_schema_version("pacemaker-4.0") <= version)) {
```
Comment (kgaillot) [inline lib/common/acl.c]: Let's define a constant for the upper bound

File: tools/cibadmin.c [hunk 13]
```diff
@@ -458,6 +536,46 @@ main(int argc, char **argv)
...
+                    crm_exit(CRM_EX_CONFIG);
+                }
+                output = NULL;
+            } else {
+                output = xmlDocGetRootElement(acl_evaled_doc);
+            }
+        } else if (rc == 0) {
+
+        } else if (rc < 0) {
+            crm_crit("Could not evaluated ACLs per request (%s, rc=%d)",
```
Comment (kgaillot) [inline tools/cibadmin.c]: spelling: evaluate

Tool errors should go to stderr rather than the log

File: lib/common/acl.c (hunk 12 above)
Comment (jnpkrn) [reply lib/common/acl.c]: For a future unfolding from users-only view.

**Comment by jnpkrn** [comment]
Was a PoC originally, but got fully baked as it was quite a furious
struggle to put the public API nits of undocumented libxslt
introspection (and manipulation) together, assisted with XML catalog
technique we already had experience with from the build system to
workaround some Pulibcan quirks.

May immediate benefit for us: the feature in cibadmin readily workable
from checkout and subsequent build (i.e., in developmental mode).

Will get back to the original feedback, yet.

File: include/crm/common/xml_internal.h (hunk 9 above)
Comment (jnpkrn) [reply include/crm/common/xml_internal.h]: this is not a typo, it's a fully valid spelling

File: lib/common/acl.c [hunk 14]
```diff
@@ -1081,6 +1084,89 @@ parse_params(xsltTransformContext *transform_ctxt, const char **fallback,
...
+    CRM_ASSERT(dname != NULL);
+
+    prev_umask = umask(S_IRUSR | S_IWUSR);
+    /* XXX flock(2) where possible? */
+    fd = mkstemp(catalog_path);
+    umask(prev_umask);
+
+    if ((f = fdopen(fd, "w")) == NULL) {
+        free(catalog_path);
+        return NULL;
```
Comment (kgaillot) [inline lib/common/acl.c]: also free dname

File: lib/common/acl.c [hunk 15]
```diff
@@ -1081,6 +1084,89 @@ parse_params(xsltTransformContext *transform_ctxt, const char **fallback,
...
+    "  <rewriteURI uriStartString='http://clusterlabs.org/nsredir/pacemaker/cfg' rewritePrefix='file://",
+            dname,
+      "'/>\n"
+      "'<nextCatalog catalog='file:///etc/xml/catalog'/>\n"
+    "</catalog>");
+
+    fclose(f);
+
+    if (free_source) {
+        free((char *) appointed_cfg);
```
Comment (kgaillot) [inline lib/common/acl.c]: I'm not a fan of breaking const signatures; it can lead to surprises and confusion. It would be easier to follow if the caller freed the value when needed (or if it was char* and free_source was the only behavior).

File: tools/cibadmin.c [hunk 16]
```diff
@@ -119,6 +131,9 @@ static struct crm_option long_options[] = {
     {"-spacer-",    0, 0, '-', " $EDITOR $HOME/local.xml", pcmk_option_example},
     {"-spacer-",    0, 0, '-', " cibadmin --replace --xml-file $HOME/local.xml", pcmk_option_example},
 
+    {"-spacer-",    0, 0, '-', "Render configuration in color per respective ACLs as evaluated for user tony:", pcmk_option_paragraph},
+    {"-spacer-",    0, 0, '-', " cibadmin -L -U tony", pcmk_option_example},
+
```
Comment (kgaillot) [inline tools/cibadmin.c]: This needs to be guarded by ENABLE_ACL as well

File: tools/cibadmin.c [hunk 17]
```diff
@@ -369,7 +424,45 @@ main(int argc, char **argv)
...
+            fflush(stderr);
+            crm_exit(CRM_EX_USAGE);
+        }
+
+        /* we already stopped/warned ACL-controlled users about consequences */
+        acl_cred = cib_user;
+        cib_user = NULL;  /* XXX CRM_DAEMON_USER as it's IPC guarded anyway? */
+        /* XXX should likely differenciate per admin_input_* */
     }
+#endif
```
Comment (kgaillot) [inline tools/cibadmin.c]: The closing brace needs to be outside the guard

Comment (jnpkrn) [reply tools/cibadmin.c]: Good catch

File: tools/cibadmin.c (hunk 16 above)
Comment (jnpkrn) [reply tools/cibadmin.c]: Oh yeah

File: lib/common/acl.c (hunk 14 above)
Comment (jnpkrn) [reply lib/common/acl.c]: yep, thanks

File: Makefile.am [hunk 18]
```diff
@@ -18,7 +18,8 @@ CORE_INSTALL	= replace include lib daemons tools xml
 # Only these will get built with a plain "make" or "make clean"
 CORE		= $(CORE_INSTALL) cts
 
-SUBDIRS	= $(CORE) doc extra maint
+# this directory needs priority treatment (install creates a relied upon dir)
+SUBDIRS	= . $(CORE) doc extra maint
```
[the remark this answers is not in the document]
Comment (jnpkrn) [reply Makefile.am]: `.` ~ where that `Makefile.am` is present ~ tree root

File: lib/common/schemas.c [hunk 19]
```diff
@@ -366,9 +367,46 @@ add_schema_by_version(const schema_version_t *version, int next,
...
+    int ret = 0;
+
+    /* security framework preferences */
+    if (!finalize) {
+        CRM_ASSERT(secprefs == NULL);
+        secprefs = xsltNewSecurityPrefs();
+        ret = xsltSetSecurityPrefs(secprefs, XSLT_SECPREF_WRITE_FILE,
+                                   xsltSecurityForbid)
+              | xsltSetSecurityPrefs(secprefs, XSLT_SECPREF_WRITE_FILE,
+                                     xsltSecurityForbid)
```
[the remark this answers is not in the document]
Comment (jnpkrn) [reply lib/common/schemas.c]: Thanks, for apparently reserved for ` XSLT_SECPREF_CREATE_DIRECTORY`.

File: tools/cibadmin.c (hunk 13 above)
Comment (jnpkrn) [reply tools/cibadmin.c]: That's the matter of already discussed terrible discrepancy in using
progress-info means ambiguously.  In an ideal world, `crm_log_init`
would use the "is daemon" flag and continue to forward towards syslog,
file logs, journal or whatever else is configured for the daemons to
use.  For end interaction programs, it would either forward it to
`stderr` or to (perhaps later on) to a custom logging handler specific
to the program in question.

It would reolve all these nasty discrepancies once forever.
End result: relaxed programmers since they use a single familiar
interface carring just about single metadata item to be attached
(severity) and leaving everything else to the "middleware" to get
the routing right.

It's sad we are not there and there were even objections to that
natural model to be adopted.

Will use nasty explicit `fprintf` instead.

File: pacemaker.spec.in [hunk 20]
```diff
@@ -642,6 +645,12 @@ exit 0
 %postun cluster-libs -p /sbin/ldconfig
 %endif
 
+%post schemas
+xmlcatalog --noout --add delegateURI \
+  http://clusterlabs.org/nsredir/pacemaker/ \
+  file://%{_datadir}/pacemaker/catalog.xml \
+  %{_sysconfdir}/xml/catalog
```
[the remark this answers is not in the document]
Comment (jnpkrn) [reply pacemaker.spec.in]: Indeed, I forgot about removal on uninstall, thanks.

Regarding the path, situation really varies across Fedora packages,
will take a closer look, but don't think it should be a critical point here.

Regarding guidelines, for long term, I've opened an idea of adopting
the openSUSE style in Fedora:
https://lists.fedoraproject.org/archives/list/packaging@lists.fedoraproject.org/message/5WP5NAC23XX6CBCDRHZVDA4BAHBEKCXH/

Picky downstream will need to adjust, and the more generic approach
will still work there, regardless.

Let me also think about `make install` and `make uninstall` repercussions,
since that's also a valid way to have the system installed.

[the remark this answers is not in the document]
Comment (jnpkrn) [reply pacemaker.spec.in]: Ok, I think it's sufficiently resolved.

OpenSUSE is expected to pass `--with-detachedxmlcatalogdir=/etc/xml/catalog.d`
and run their devised commands in `%post` and `%postun`.

File: tools/cibadmin.c [hunk 21]
```diff
@@ -313,19 +317,19 @@ main(int argc, char **argv)
...
             default:
                 printf("Argument code 0%o (%c)" " is not (?yet?) supported\n", flag, flag);
                 ++argerr;
                 break;
         }
+        if (cib_action_old != NULL && cib_action != cib_action_old) {
+            printf("Combining multiple commands into one go not supported\n");
+            ++argerr;
+	    break;
+        }
```
[the remark this answers is not in the document]
Comment (jnpkrn) [reply tools/cibadmin.c]: How does this prevent specifying such aliases?  If course, if
someone unreasonably relied upon undefined behaviour, they are going,
deservedly, to be bitten.  But nothing unexpected, you simply don't
do stupid things like that.

Definitely better then allowing proliferation of ill-advised command
combinations.  Do you know of any command that would allow combining
action triggers like that, relying on some undocumented decision rule
about what to actually trigger?  It's simply unsupportable and better
to indicate that clearly, to prevent bad surprises (that it wasn't
the case in the past is already bad enough).

[the remark this answers is not in the document]
Comment (jnpkrn) [reply tools/cibadmin.c]: It also prevents command injection should any of the official or
home-made higher level management tools be naive enough, so you can
take it also a very desired security hardening -- you don't want
anyone be able to completely shift the semantics of the command
to execute like that.

File: lib/common/acl.c (hunk 15 above)
Comment (jnpkrn) [reply lib/common/acl.c]: Agreed, and luckily, made me look deeper into `libxml2/catalog.c`,
leading to realization that I am going the bad route -- it doesn't
matter that much in one-off execution, but spoiling global state
of 3rd party library is wrong anyway (consider the static function
bubbling up to something widely reused, also in daemon context,
for instance).  So this whole chunk is getting an overhaul.

**Comment by jnpkrn** [comment]
I attempted to avoid any global state mangling and per-document
(or rathar, per parser context) catalogs were very promissing.
Unfortunately, it was a dead end, for the last "millimeter" piece
in chain missing (`xsltDocDefaultLoaderFunc` not adopting existing
context but creating a default, non-customized one anew, even if
the context object gets passed into it!).

So the global state needs to be modified regardless.  Externalized
catalog file and mangling with per-process globally visible environment
variable is hefty and error-prone, so eventually the solution is to
install a custom document loader that will intercept and rewrite
particular URI on the fly, then deferring to the previously installed
(presumably default) document loader.

File: lib/common/acl.c (hunk 12 above)
Comment (jnpkrn) [reply lib/common/acl.c]: That was meant as a response to `PCMK_ACL_CRED_USER`.
This point is resolved now.

**Comment by jnpkrn** [comment]
Travis-CI, `MAINT=1`:

> Relax-NG validity error : Extra element choice in interleave
> pacemaker-acls-3.0.rng:10: element choice: Relax-NG validity error : Element attribute failed to validate content
> pacemaker-acls-3.0.rng:9: element optional: Relax-NG validity error : Invalid sequence in interleave
> pacemaker-acls-3.0.rng:9: element optional: Relax-NG validity error : Element optional failed to validate content
> pacemaker-acls-3.0.rng:8: element interleave: Relax-NG validity error : Invalid sequence in interleave
> Relax-NG validity error : Extra element define in interleave
> pacemaker-acls-3.0.rng:5: element define: Relax-NG validity error : Element grammar failed to validate content
> pacemaker-acls-3.0.rng fails to validate

This is exactly the manifestation of libxml2 bug identified, fixed, and
hence eventually going away:

  https://gitlab.gnome.org/GNOME/libxml2/merge_requests/31

Will silence this out for the time being.

File: lib/cib/cib_ops.c [hunk 22]
```diff
@@ -443,6 +443,11 @@ update_cib_object(xmlNode * parent, xmlNode * update)
 
     copy_in_properties(target, update);
 
+    if (xml_acl_denied(target)) {
+        crm_notice("Cannot update <%s id=%s>", crm_str(object_name), crm_str(object_id));
```
Comment (kgaillot) [inline lib/cib/cib_ops.c]: Let's improve the user-friendliness of this message, something like:

     Cannot update <%s%s%s>: Access denied by ACLs
     object_name, (object_id? " id=" : ""), (object_id? object_id : "")

(object_name is guaranteed non-NULL at this point)

File: lib/common/acl.c [hunk 23]
```diff
@@ -154,6 +154,7 @@ __xml_acl_create(xmlNode *xml, GList *acls, enum xml_private_flags mode)
  *
  * \return New head of (possibly modified) acls
  */
+PCMK__XML_DIRECTLY_RECURSIVE
```
Comment (kgaillot) [inline lib/common/acl.c]: This requires someone reading to code to go grep for this constant. I think it makes more sense to just add "\note This function is recursive" in the doxygen block, where it will be useful for API readers as well.

File: include/crm/compatibility.h [hunk 24]
```diff
@@ -226,6 +226,18 @@ get_resource_typename(enum pe_obj_types type)
...
+ * Version compatibility tracking incl. open-ended intervals for occasional
+ * bumps (to avoid hard to follow open-coding throughout).  Grouped by context.
+ */
+
+/* Schema version vs. evaluate-as-namespace-annotations-per-credentials */
+
+#define PCMK_COMPAT_ACL_2_MIN_INCL "pacemaker-2.0"
+#define PCMK_COMPAT_ACL_2_MAX_EXCL "pacemaker-4.0"  /* bump if remains OK */
+
+
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply include/crm/compatibility.h]: I've added 379f24e7 to make the purpose of compatibility.h clearer

**Comment by kgaillot** [comment]
FYI, I cherry-picked the "restrictive usage", "serializer reinventing", and "get_schema_path" commits into #1914 so they get exercised in 2.0.3-rc1. The remainder of this can remain against master branch.

**Comment by jnpkrn** [comment]
Rebased on top of #1943 with some tweaks and fixes.

Most importantly, I decided to refrain from `acls` in parts directly
facing the user: the switch is now `--access-render`, which is
substantially more user friendly, as it touches the merit of the
action directly (rather than via obnoxious TLA, which is moreover
not a proper fit here as documented).

To immediately fix:

- superfluous, directory specific message in `./cts-cli access_render`

- `acl-2-render.xsl` doesn't restore the color after returning from tag
  correctly in all cases (as attested with said test suite easily)

**Comment by jnpkrn** [comment]
Mentioned and couple of more issues were fixed and some more
tweaks and rectifications absorbed.

File: tools/cibadmin.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (jnpkrn) [reply tools/cibadmin.c]: Yes, exactly because it's not a good idea to surprise users,
it's better not to bet they will never forget extra dash as described.
I don't see a value in extra complexity only to be removed one sweep
later, do you?

File: Makefile.am (hunk 18 above)
[the remark this answers is not in the document]
Comment (jnpkrn) [reply Makefile.am]: Doesn't explain what's wrong with the current approach, does it?

You likely meant `make -C xml install`, but that's rather unsupported by
definition (you want project installed, not it's part like this; the craziness
around this microoptimizing would be avoided with this problems causing
recursive approach altogether ... #1701).

File: tools/cibadmin.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (jnpkrn) [reply tools/cibadmin.c]: Especially when higher level tools that may perhaps call down to this
are somewhat expected to construct every such command in a sane
way, meaning they shall not be influenced a single bit.

Verging on the undefined behaviour is always unwise.

That works both ways, we are not doing a good job of positively
defining what exactly is this or other option/switch/subcommand
meant to trigger (see recent discussions, `--text-fancy`, what
are stability-over-time expectations -- they are actually more
important than whether a mix of actions, something bad from
beginning, is recognized deterministically over time).

File: include/crm/common/xml.h (hunk 10 above)
Comment (jnpkrn) [reply include/crm/common/xml.h]: My view radically differs, this is highly practical for simple projects
for which actually any bloated build system (like said `configure` checks)
is a clutter on it's own, when they can just `#ifdef` in code as they see
fit -- conclusion shall not be jumped to based on a single project.

Calling 3 extra lines a clutter?  This is the same approach used
in the standard library, one would be mad translating any ad-hoc
`#ifdef`s in code into `configure` checks + consequently another
`#ifdef`s based on the result of those `configure` checks (doesn't
it resemble the notion of "clutter"?).

Also note that `configure` checks can never be as up-to-date as `#ifdef`
when actually compiling (check-during-build is sort of atomic and the
only _static_ source of truth, and regarding _static_ limitations, that's
why there's this further _dynamic_ indicated-unsupportability-by-return-code
rectification when the two accidentally differ).

**Comment by kgaillot** [comment]
This has been forward-ported to the current code base over several PRs
